strawberryfields/engine.py

```diff
@@ -603,21 +604,64 @@ def run_async(self, program: Program, *, compile_options=None, **kwargs) -> Job:
         Keyword Args:
             shots (Optional[int]): The number of shots for which to run the job. If this
                 argument is not provided, the shots are derived from the given ``program``.
+            recompile (bool): specifies if ``program`` should be recompiled
+                using the default compilation options
 
         Returns:
             strawberryfields.api.Job: the created remote job
         """
         # get the specific chip to submit the program to
         compile_options = compile_options or {}
+        recompile = kwargs.pop("recompile", False)
         kwargs.update(self._backend_options)
 
         device = self.device_spec
 
         compiler_name = compile_options.get("compiler", device.default_compiler)
-        msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
-        self.log.info(msg)
 
-        program = program.compile(device=device, **compile_options)
+        program_is_compiled = program.compile_info is not None
+
+        if program_is_compiled and not recompile:
+            # error handling for program compilation:
+            # program was compiled but recompilation was not allowed by the
+            # user
+
+            if (
+                program.compile_info[0].target != device.target
+                or program.compile_info[0]._spec != device._spec
```

Comment:
Did we reach a conclusion regarding this? I guess it doesn't harm to keep both checks, although I still feel like just checking `device._spec` should be enough, and then update `DeviceSpec` by either removing `target` or adding a warning/error when creating using a different target than the one in the layout.

Reply:
Thanks for raising the point again! Created https://github.com/XanaduAI/strawberryfields/issues/449 as per @josh146's suggestion to mark it.

From my perspective, if there's no reason against having both checks, I'd rather keep them for now. The reason is just super safety in exchange for minor overhead.

Reply:
Yep, I don't think there is a reason for now against having both checks. It just looks a bit surprising, and indicates potential inconsistency earlier in the stack that should be double checked!